cocos/serialization/compiled/builtin-value-type.ts

```diff
@@ -0,0 +1,129 @@
+import { JSB } from 'internal:constants';
+import { Vec2, Vec3, Vec4, Color, Size, Rect, Quat, Mat4, assertIsTrue } from '../../core';
+import type { IIntrudedFileData } from '../deserialize';
+
+const constructorMap = [
+    Vec2,   // 0
+    Vec3,   // 1
+    Vec4,   // 2
+    Quat,   // 3
+    Color,  // 4
+    Size,   // 5
+    Rect,   // 6
+    Mat4,   // 7
+] as const;
+
+type ConstructorMap = typeof constructorMap;
+
+export type ValueTypeData = [
+    typeIndex: number,
+    ...values: number[],
+];
+
+type SetterInput = [reserved: number, ...values: number[]];
+
+type Setter<T> = (value: T, result: SetterInput) => void;
+
+type SetterMapImpl<T extends readonly any[]> = T extends readonly [infer Head, ...infer Tails]
+    ? Head extends abstract new (...args: any) => any
+        ? readonly [Setter<InstanceType<Head>>, ...SetterMapImpl<Tails>]
+        : readonly [Head, ...SetterMapImpl<Tails>]
+    : readonly [];
+
+type SetterMap = SetterMapImpl<ConstructorMap>;
+
+function vec4LikeSetter<T extends Vec4 | Quat> (obj: T, data: SetterInput): void {
+    obj.x = data[1];
+    obj.y = data[2];
+    obj.z = data[3];
+    obj.w = data[4];
+}
+
+const setterMap: SetterMap = [
+    (obj: Vec2, data: SetterInput): void => {
+        obj.x = data[1];
+        obj.y = data[2];
+    },
+    (obj: Vec3, data: SetterInput): void => {
+        obj.x = data[1];
+        obj.y = data[2];
+        obj.z = data[3];
+    },
+    vec4LikeSetter,   // Vec4
+    vec4LikeSetter,   // Quat
+    (obj: Color, data: SetterInput): void => {
+        obj._val = data[1];
+    },
+    (obj: Size, data: SetterInput): void => {
+        obj.width = data[1];
+        obj.height = data[2];
+    },
+    (obj: Rect, data: SetterInput): void => {
+        obj.x = data[1];
+        obj.y = data[2];
+        obj.width = data[3];
+        obj.height = data[4];
+    },
+    (obj: Mat4, data: SetterInput): void => {
+        Mat4.fromArray(obj, data, 1);
+    },
+];
+
+export function serializeBuiltinValueType (obj: unknown): ValueTypeData | undefined {
+    if (typeof obj !== 'object' || !obj) {
```

Comment:
why not declare obj as ValueType? In that case, the judgment is redundant

Reply:
Solved: https://github.com/cocos/cocos-engine/pull/15892/commits/6a24bc01aa0501569a04715dd650902cd7d52ff2